File: java/2/Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(0);
        ListNode p = head;
        int carry = 0;
        while(l1!=null || l2!=null){
            int sum = carry;
            if(l1!=null){
                sum += l1.val;
                l1 = l1.next;
            }
            if(l2!=null){
                sum += l2.val;
                l2 = l2.next;
            }
            carry = sum/10;
            sum %= 10;
            p.next = new ListNode(sum);
            p = p.next;
        }
        if(carry>0){
            p.next = new ListNode(carry);
        }
        return head.next;
    }
}

File: cpp/2/Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = new ListNode(0);
        ListNode *p = head;
        int carry = 0;
        while(l1 || l2 || carry>0){
            carry += getValue(l1)+getValue(l2);
            p->next = new ListNode(carry%10);
            p = p->next;
            l1 = getNext(l1);
            l2 = getNext(l2);
            carry /= 10;
        }
        p = head->next;
        delete head;
        return p;
    }
    
    int getValue(ListNode *p){
        return p==NULL?0:p->val;
    }
    
    ListNode* getNext(ListNode *p){
        return p==NULL?NULL:p->next;
    }
};